<?xml version="1.0"?>
<robot name="diff_drive_robot" xmlns:xacro="http://ros.org/wiki/xacro">
  <!-- Константы -->
  <xacro:property name="base_length" value="1"/>
  <xacro:property name="base_width" value="1"/>
  <xacro:property name="base_height" value="0.5"/>
  <xacro:property name="wheel_radius" value="0.3"/>
  <xacro:property name="wheel_length" value="0.2"/>
  <xacro:property name="wheel_mass" value="0.2"/>
  <xacro:property name="base_mass" value="1.0"/>

  <link name="world"/>

  <joint name="world_to_base" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <!-- Базовый корпус робота -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="${base_length} ${base_width} ${base_height}"/>
      </geometry>
      <material name="Gazebo/Blue"/>
    </visual>
    <collision>
      <geometry>
        <box size="${base_length} ${base_width} ${base_height}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${base_mass}"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
  </link>

  <!-- Левое заднее колесо -->
  <link name="left_back_wheel">
    <visual>
      <geometry>
        <cylinder length="${wheel_length}" radius="${wheel_radius}"/>
      </geometry>
      <material name="Gazebo/Blue"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="${wheel_length}" radius="${wheel_radius}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${wheel_mass}"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
  </link>

  <!-- Правое заднее колесо -->
  <link name="right_back_wheel">
    <visual>
      <geometry>
        <cylinder length="${wheel_length}" radius="${wheel_radius}"/>
      </geometry>
      <material name="Gazebo/Blue"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="${wheel_length}" radius="${wheel_radius}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${wheel_mass}"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
  </link>

  <!-- Соединения для колес -->
  <joint name="left_back_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_back_wheel"/>
    <!--origin xyz="${base_length/-2 - wheel_length/2} ${base_width/2} 0" rpy="0 ${pi/2*1} 0"/-->
    <origin xyz="${base_length/-2 - wheel_length/2} 0 0" rpy="0 ${pi/2*1} 0"/>
    <axis xyz="0 0 1"/>
  </joint>

  <joint name="right_back_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_back_wheel"/>
    <!--origin xyz="${base_length/2 + wheel_length/2} ${base_width/2} 0" rpy="0 ${pi/2*1} 0"/-->
    <origin xyz="${base_length/2 + wheel_length/2} 0 0" rpy="0 ${pi/2*1} 0"/>
    <axis xyz="0 0 1"/>
  </joint>

  <!-- Плагин для публикации состояний соединений -->
  <gazebo>
    <plugin name="gazebo_ros_joint_state_publisher" filename="libgazebo_ros_joint_state_publisher.so">
      <ros>
        <namespace>/</namespace>
      </ros>
      <joint_name>left_back_wheel_joint</joint_name>
      <joint_name>right_back_wheel_joint</joint_name>
      <!--joint_name>left_front_wheel_joint</joint_name-->
      <!--joint_name>right_front_wheel_joint</joint_name-->
    </plugin>
  </gazebo>

  <!-- Плагин для дифференциального привода -->
  <gazebo>
    <plugin name="gazebo_ros_diff_drive" filename="libgazebo_ros_diff_drive.so">
      <ros>
        <namespace>/</namespace>
        <remapping>cmd_vel:=/cmd_vel</remapping> <!-- Подключение к топику /cmd_vel -->
      </ros>
      <update_rate>100.0</update_rate> <!-- Частота обновления -->
      
      <left_joint>left_back_wheel_joint</left_joint> <!-- Левое колесо -->
      <right_joint>right_back_wheel_joint</right_joint> <!-- Правое колесо -->
      <!--left_joint>left_front_wheel_joint</left_joint--> <!-- Левое колесо -->
      <!--right_joint>right_front_wheel_joint</right_joint--> <!-- Правое колесо -->
      
      <wheel_separation>${base_width + wheel_length}</wheel_separation> <!-- Расстояние между колесами -->
      <wheel_diameter>${wheel_radius*2}</wheel_diameter> <!-- Радиус колес -->

      <odometrySource>joint_states</odometrySource> <!-- Источник одометрии -->
    </plugin>
  </gazebo>

    <!-- Лидар -->
  <link name="lidar_link">
    <visual>
      <geometry>
        <cylinder length="0.1" radius="0.05"/>
      </geometry>
      <material name="Gazebo/Blue"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.1" radius="0.05"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial  > 
  </link>

  <joint name="lidar_joint" type="fixed">
    <parent link="base_link"/>
    <child link="lidar_link"/>
    <origin xyz="0 0 0.5" rpy="0 0 0"/> 
  </joint>

  <gazebo reference="lidar_link">
    <sensor type="gpu_lidar" name="gpu_lidar">
      <update_rate>10</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>360</samples>
            <resolution>1</resolution>
            <min_angle>-1.57</min_angle>
            <max_angle>1.57</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.2</min>
          <max>10.0</max>
          <resolution>0.01</resolution>
        </range>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.01</stddev>
        </noise>
      </ray>
      <plugin name="gazebo_ros_laser" filename="libgazebo_ros_gpu_laser.so">
        <ros>
          <namespace>/</namespace>
        </ros>
        <topicName>/scan</topicName>
        <frameName>lidar_link</frameName>
      </plugin>
    </sensor>
  </gazebo>
</robot>
